spanner/session.go

```diff
@@ -1107,6 +1208,79 @@ func (p *sessionPool) take(ctx context.Context) (*sessionHandle, error) {
 	}
 }
 
+// takeMultiplexed returns a cached session if there is available one; if there isn't
+// any, it tries to allocate a new one.
+func (p *sessionPool) takeMultiplexed(ctx context.Context) (*sessionHandle, error) {
+	if !p.enableMultiplexSession {
+		return p.take(ctx)
+	}
+	trace.TracePrintf(ctx, nil, "Acquiring a multiplexed session")
+	for {
+		var s *session
+
+		p.mu.Lock()
+		if !p.valid {
+			p.mu.Unlock()
+			return nil, errInvalidSessionPool
+		}
+		// use the multiplex session if it is available
+		if p.multiplexedSession != nil {
+			// Multiplexed session is available, get it.
+			s = p.multiplexedSession
+			trace.TracePrintf(ctx, map[string]interface{}{"sessionID": s.getID()},
+				"Acquired multiplexed session")
+		}
+		if s != nil {
+			p.mu.Unlock()
+			// From here, If healthcheck workers failed to
+			// schedule healthcheck for the session timely, do the check here.
+			// Because session check is still much cheaper than session
+			// creation, they should be reused as much as possible.
+			if !p.isHealthy(s) {
+				continue
+			}
+			p.incNumMultiplexedInUse(ctx)
+
+			return p.newSessionHandle(s), nil
+		}
+
+		// No session available. Start the creation of multiplexed session.
+		if err := p.getMultiplexedSession(ctx); err != nil {
+			p.mu.Unlock()
+			return nil, err
+		}
+
+		mayGetSession := p.mayGetMultiplexedSession
```

Comment:
This point will be reached when the session pool has no multiplexed session. However, what happens when there are for example 2 goroutines that both try to get a session when there is no multiplexed session in the pool is the following:
1. The first goroutine tries to get a session. There is none, and the creation of a new one is initiated. The goroutine waits here.
2. The second goroutine tries to get a session. There is still none, and the creation of a new one is initiated. The goroutine waits here.
3. One of the multiplexed sessions that was created by either of the above goroutines is ready and released into the pool. Both the above goroutines will use that session.
4. The other multiplexed session that was created by the two above goroutines is ready and released into the pool. This will now become the multiplexed session of this pool.

In other words: This implementation creates more multiplexed sessions than necessary. The above example is for 2 goroutines, but if there are 100, then up to 100 could be created.

Reply:
There is a single listener for events from multiplexedSessionReq so at any given point of time only 1 session creation will be in process